### ISSUE TITLE
value of skipped is "1" in testng-results.xml when a test fails and retried to execute.

### ISSUE THREAD

**Issue by praveen133t** [testng-team/testng#878]
Created a simple test with Assert.fail() and retried the same test using a class that implements IRetryAnalyzer. In the testng-results.xml the "skipped" attribute has value "1" and "failed" attribute has value "1". This is incorrect as there are no tests skipped. It's working fine in 6.9.4 version.

```
public class SkipFailTest {

    @Test(retryAnalyzer=RetryAnalyzer.class)
    public void testSkipFailStatus() {
        Assert.fail();
    }
}

    @Override
    public boolean retry(ITestResult result) {
        if (!result.isSuccess()) {
            if (count < maxCount) {
                count++;
                return true;
            }
        }
        return false;
    }
```

testng-results.xml has `<testng-results skipped="1" failed="1" total="2" passed="0">`

The details of the test also has "SKIP" as test "status"

**Comment by praveen133t** [comment]
@juherr : Quickly looking at the code I see in the Invoker class in the method `handleInvocationResults` there is a logic inside the `if` loop 

      if (willRetry) {
        resultsToRetry.add(testResult);
        failure.count++;
        failure.instances.add(testResult.getInstance());
        testResult.setStatus(ITestResult.SKIP);
      }
Why are we doing testResult.setStatus(ITestResult.SKIP); inside retry? If that line is commented out the results works fine.

**Comment by juherr** [comment]
@rschmitt should be able to answer :) https://github.com/cbeust/testng/blame/master/src/main/java/org/testng/internal/Invoker.java#L1391-L1395

**Comment by rschmitt** [comment]
Of all the *existing* statuses, `ITestResult.SKIP` seemed like the best fit for "test failed, pending retry." It's arguably far more correct than a status of `PASS`. Adding a whole new test result status seemed like it would be a very invasive change, given all the IDE plugins and so forth that are out there, and my immediate goal was to fix the much more severe `IRetryAnalyzer` issues involving exit codes and so forth.

**Comment by juherr** [comment]
@rschmitt But how did it work in 6.9.4? And what about the `ITestResult.STARTED` status instead?

**Comment by praveen133t** [comment]
@juherr  : I might be wrong about this issue working fine in 6.9.4. Because I remove the result of the first run (before retry) using

ITestContext context = result.getTestContext();
				IResultMap toRemoveTest = context.getFailedTests();
				List<ITestNGMethod> tcToRemove = new ArrayList<ITestNGMethod>();
				tcToRemove.add(result.getMethod());			
				for(ITestNGMethod failedTCToRemove : tcToRemove)
				{
					toRemoveTest.removeResult(failedTCToRemove);
				}

The above code doesn't work any more. I looked into that a little bit as well. Looks like `m_map` is empty in the `removeResult` method.

This was working fine in 6.9.4. Since this broke in 6.9.9 I noticed the SKIP issue.

**Comment by rschmitt** [comment]
@juherr In 6.9.4 the test would be marked as `FAILED`. This failure status would propagate to various listeners, most importantly the `ExitCodeListener`, which would mark the entire test suite as failed by emitting a non-zero exit code. As far as I can tell, @praveen133t was previously working around this bug by using a listener that would "sanitize" the test results by mutating the `result` argument, a practice which has undefined behavior.

**Comment by juherr** [comment]
@praveen133t Ok, so as `FAILED` was replaced by `SKIPPED`, it should work if you replace `getFailedTests` by `getSkippedTests`. Or you can just remove this code if you tried to fix the problem described by @rschmitt.

@rschmitt And what do you think about `STARTED` instead of `SKIPPED`? I have no idea about side effects for the moment but a very quick hack showed me only 2 regressions in tests.

**Comment by rschmitt** [comment]
@juherr What problem are you trying to solve? Why is `STARTED` better than `SKIPPED`?

**Comment by juherr** [comment]
@rschmitt The "problem" is `ITestListener#onTestSkipped()` is notified and `ITestListener#onTestFailure()` was notified before.
I'm just asking if `ITestListener#onTestStart()` or even none could be better.

For example, see: https://github.com/juherr/testng/blob/issue-857/src%2Ftest%2Fjava%2Ftest%2FretryAnalyzer%2FRetryAnalyzerTest.java#L50-L51
(strange behavior too: `tla.getPassedTests()` is empty)

What do you think?
ping @cbeust

**Comment by praveen133t** [comment]
@juherr : I even tried doing `getSkippedTests` to remove the `skipped` result from the 'testng-results.xml' but even that doesn't work. I still see the `SKIPPED` result in the xml. Pls [see this.] (http://pastebin.com/tEag8KB2) I think that's a whole new problem? As far as I can see `m_map` is not getting populated with the test results.

**Comment by juherr** [comment]
@praveen133t Ok, I take the point but I won't be able to look deeper soon.

**Comment by praveen133t** [comment]
@juherr : Sure no problem. Do you want me to create a new issue for the `removeResult`? I can take a look at the issue and post something useful.

**Comment by juherr** [comment]
@praveen133t Yes, you can. And if you can share a test case which shows the issue in a pr, it will be welcome.

**Comment by cbeust** [comment]
I agree that listeners should probably not be notified as long as there retry analyzers working on a test but I'm also not a fan of adding a new status (which has a lot of repercussions, as @rschmitt pointed out).

**Comment by MichaelKunze** [comment]
I came across this issue. Personally i really would prefer a new state (eg: "RETRIED"). 
Neither of the established states seems to reflect the proper state.
PASSED? obviously not.
SKIPPED? Well... Test was executed, wasn't it? Then one would assume: FAILED. But we all know that triggers all the ExitCodesListener and makes things worse.

Any thoughts?

**Comment by derrdji** [comment]
Without reading the source code, I would imagine adding a new "RETRIED" state as @MichaelKunze suggested might make fixing this bug easier; because there will be no confusion of which "STARTED" it really is.

The same bug (or a different one) is also causing the Retry Analyzer being called again with the new result status.

I have a simple test `allFailTest()` that does `Assert.fail();`, I have a test listener counting the number of `onTestStart()` calls to save the number of invocation (notice number of calls to `retry()` is not reliable as I will demonstrate), I have a SimpleRetryAnalyzer that keep retry until total number of execution is 3 (2 retries). I have pasted the debug prints in the Listener and the Anazlyzer. 

```
onTestStart of RetryThenFailTestClass.allFailTest()
Registering test method RetryThenFailTestClass.allFailTest() current exe number is 1
SimpleRetryAnalyzer [SimpleRetryAnalyzer@7113b13f] deciding for test RetryThenFailTestClass.allFailTest() with status = Failure, total exe count = 1
onTestSkipped of RetryThenFailTestClass.allFailTest()
onTestStart of RetryThenFailTestClass.allFailTest()
Registering test method RetryThenFailTestClass.allFailTest() current exe number is 2
SimpleRetryAnalyzer [SimpleRetryAnalyzer@7113b13f] deciding for test RetryThenFailTestClass.allFailTest() with status = Failure, total exe count = 2
onTestSkipped of RetryThenFailTestClass.allFailTest()
SimpleRetryAnalyzer [SimpleRetryAnalyzer@7113b13f] deciding for test RetryThenFailTestClass.allFailTest() with status = Skip, total exe count = 2
onTestStart of RetryThenFailTestClass.allFailTest()
Registering test method RetryThenFailTestClass.allFailTest() current exe number is 3
SimpleRetryAnalyzer [SimpleRetryAnalyzer@7113b13f] deciding for test RetryThenFailTestClass.allFailTest() with status = Failure, total exe count = 3
onTestFailedButWithinSuccessPercentage of RetryThenFailTestClass.allFailTest()
SimpleRetryAnalyzer [SimpleRetryAnalyzer@7113b13f] deciding for test RetryThenFailTestClass.allFailTest() with status = SUCCESS_PERCENTAGE_FAILURE, total exe count = 3
SimpleRetryAnalyzer [com.puppetmaster.testng.SimpleRetryAnalyzer@7113b13f] deciding for test RetryThenFailTestClass.allFailTest() with status = SUCCESS_PERCENTAGE_FAILURE, total exe count = 3
SKIPPED: allFailTest
...
```

Things I observed and do not understand:

1. `SimpleRetryAnalyzer.retry()` is invoked more times than `Listener.onTestStart()`
2. After the 1st execution failure, the result was changed to `SKIPPED`, and `onTestSkipped()` was invoked (also confirmed by @praveen133t). After the 2nd execution failure, the result was changed to `SKIPPED`, `onTestSkipped()` was invoked, and the `Analyzer.retry()` was invoked again with result status being Skipped. Notice this was not the case after the 1st failure.
3. After the 3rd failure, the result status was FAILURE, but the listener's `onTestFailedButWithinSuccessPercentage` method was invoked. After that, the retry analyzer's `retry() `method was invoked twice with result status `SUCCESS_PERCENTAGE_FAILURE`

**Comment by praveen133t** [comment]
Hi,

Any updates on this defect?

**Comment by juherr** [comment]
No updates, sorry. 
Feel free to try to propose a fix if you can, it will help :)

**Comment by yardening2** [comment]
Hi,

I had this problem when tests that passed on retry where marked as skipped so they where generated again on testng-failed.xml and I rerun them again.
I didn't want to rerun tests that passed when retried.

I created a class that extends TestListenerAdapter and modified onTestSuccess as so:

```java
@Override
public void onTestSuccess(ITestResult tr) {
	super.onTestSuccess(tr);
	tr.getTestContext().getSkippedTests().removeResult(tr.getMethod());
}
```

and added it as a listener.
no more problems.

How about this fix?

**Comment by praveen133t** [comment]
I see this issue has been fixed? Earlier we used to see "SKIP" as the test result for the initial test run and "FAIL" for the retry. But now I see "FAIL" and "FAIL" for both initial run and retry.

**Comment by juherr** [comment]
@praveen133t Which version did you test? @krmahadevan fixed #1519 in the current snapshot and it may explain the new result.

**Comment by juherr** [comment]
Reviewed against current retry/reporting code: the original reporting defect is resolved.

TestNG still uses `ITestResult.SKIP` internally for an attempt that will be retried, but it now records that state separately with `ITestResult.wasRetried()`. The built-in reporting path distinguishes retried attempts from genuine skips: `AbstractXmlReporter` excludes retried results from the skipped count, and the HTML/text reporters apply the same distinction. This capability was introduced by merged PR #1812 for #1697, after the retry-reporting changes discussed here.

I did not rerun the 2015 reproducer, but the current implementation directly addresses the reported `skipped` count ambiguity. Closing as resolved.
